File: LR2/7/7.rb
# Функция для подсчета стержней с кольцами всех трех цветов
def count_points(rings)
    # Инициализируем массив из 10 хэшей, каждый представляет стержень
    # Каждый хэш содержит флаги для цветов R, B, W (по умолчанию false)
    rods = Array.new(10) { { r: false, b: false, w: false } }
  
    # Проходим по строке с шагом 2 (каждые два символа - пара цвет-позиция)
    (0...rings.length).step(2) do |i|
      # Проверяем, что не выходим за пределы строки
      next if i + 1 >= rings.length
  
      color_char = rings[i].downcase # Приводим к нижнему регистру для удобства
      position_char = rings[i + 1]
  
      # Пропускаем недопустимые цвета (только R, B, W)
      next unless ['r', 'b', 'w'].include?(color_char)
  
      # Пропускаем недопустимые позиции (только цифры 0-9)
      next unless position_char =~ /\d/ # Проверка на цифру
      position = position_char.to_i
  
      # Обновляем соответствующий флаг в хэше стержня
      rods[position][color_char.to_sym] = true
    end
  
    # Считаем количество стержней, у которых все три цвета присутствуют
    rods.count { |rod| rod[:r] && rod[:b] && rod[:w] }
  end
  
  # Основная программа
  puts "Введите строку с кольцами:"
  input = gets.chomp
  
  # Защита от пустого ввода
  if input.empty?
    puts "Ошибка: введена пустая строка."
    exit
  end
  
  # Проверка на минимальную длину (хотя бы одна пара символов)
  if input.length < 2
    puts "Ошибка: строка слишком короткая."
    exit
  end
  
  # Вызываем функцию и выводим результат
  puts "Результат: #{count_points(input)}"

File: LR2/7/7.swift
// Определение структуры для хранения информации о цветах на стержне
struct Rod {
    var hasR = false 
    var hasB = false 
    var hasW = false 
}

// Функция для подсчета стержней с кольцами всех трех цветов
func countPoints(_ rings: String) -> Int {
    // Инициализация массива из 10 стержней
    var rods = Array(repeating: Rod(), count: 10)
    let characters = Array(rings) // Преобразование строки в массив символов
    let length = characters.count
    
    // Обработка пар символов 
    for i in stride(from: 0, to: length, by: 2) {
        // Проверка на выход за пределы массива (нечетная длина строки)
        guard i + 1 < length else { break }
        
        let color = characters[i]    
        let positionChar = characters[i + 1] 
        
        // Проверка, что позиция является цифрой
        guard positionChar.isNumber else { continue }
        let position = Int(String(positionChar))! // Преобразование в Int
        
        // Проверка допустимости позиции (0...9)
        guard position >= 0 && position < 10 else { continue }
        
        // Обновление флагов стержня в зависимости от цвета
        switch color {
        case "R": rods[position].hasR = true
        case "B": rods[position].hasB = true
        case "W": rods[position].hasW = true
        default: continue // Игнорирование недопустимых цветов
        }
    }
    
    // Подсчет стержней, где есть все три цвета
    return rods.filter { $0.hasR && $0.hasB && $0.hasW }.count
}

// Основная программа
print("Input: ")
if let input = readLine() { // Чтение входной строки
    print(countPoints(input)) // Вывод результата
}